server/src/main/java/org/activityinfo/server/endpoint/akvo/flow/AkvoFlowFeed.java

```diff
@@ -0,0 +1,125 @@
+import static org.activityinfo.model.resource.Resources.toJson;
+
+public class AkvoFlowFeed implements FeedService {
+    final private ResourceId parameterFormClassId;
+    final private ResourceLocatorSync locator;
+
+    public AkvoFlowFeed(ResourceId parameterFormClassId, ResourceLocatorSync locator) {
+        this.parameterFormClassId = parameterFormClassId;
+        this.locator = locator;
+    }
+
+    @Override
+    public FormClass getParameterFormClass() {
+        return locator.getFormClass(parameterFormClassId);
+    }
+
+    @Override
+    public void updateFeed(FormClass formClass, FormInstance parameters) {
+        try (AkvoFlow akvoFlow = initialize(parameters)) {
+            updateFeed(akvoFlow, formClass, parameters);
+        }
+    }
+
+    private void updateFeed(AkvoFlow akvoFlow, FormClass formClass, FormInstance parameters) {
+        List<FormInstance> formInstances = Lists.newArrayList();
+        ResourceId formClassId = formClass.getId(), timestampId = null;
+        Map<String, FormField> formFields = Maps.newHashMap();
+
+        for (FormField formField : getParameterFormClass().getFields()) {
+            if ("timestamp".equals(formField.getCode()) && parameters.getString(formField.getId()) != null) {
+                timestampId = formField.getId();
+            }
+        }
+
+        if (timestampId == null) {
+            throw new IllegalStateException("ParameterFormClass is invalid");
+        }
+
+        for (FormField formField : formClass.getFields()) {
+            formFields.put(formField.getCode(), formField);
+        }
+
+        for (SurveyInstance instance : akvoFlow.getSurveyInstances(parameters, timestampId)) {
+            FormInstance formInstance = new FormInstance(CuidAdapter.newLegacyFormInstanceId(formClassId), formClassId);
+            Date start = new Date(instance.collectionDate);
+            Date end = new Date(instance.collectionDate + instance.surveyalTime * 1000L);
+
+            formInstance.set(CuidAdapter.field(formClassId, CuidAdapter.START_DATE_FIELD), start);
+            formInstance.set(CuidAdapter.field(formClassId, CuidAdapter.END_DATE_FIELD), end);
+
+            for (QuestionAnswer questionAnswer : akvoFlow.getQuestionAnswers(instance.keyId)) {
+                FormField formField = formFields.get(questionAnswer.textualQuestionId);
+
+                if (formField == null) throw new IllegalStateException("Can't get " + questionAnswer.textualQuestionId);
+
+                for (EnumItem enumItem : ((EnumType) formField.getType()).getValues()) {
+                    if (enumItem.getLabel().equals(questionAnswer.value)) {
+                        formInstance.set(formField.getId(), new EnumValue(enumItem.getId()));
+                    }
+                }
+            }
+
+            formInstances.add(formInstance);
+        }
+
+        for (FormInstance formInstance : formInstances) {
+            getDefaultQueue().add(withUrl("/tasks/persist")
```

Comment:
I agree that a single worker task is probably better, but I'm not sure about the specifics. When faced with timeouts, it won't be easy to save a list of form instances that remain to be stored. I believe App Engine throws some kind of soft timeout exception about one second before killing a task, but catching it and then saving the current state within a single second doesn't sound like a fool-proof solution.

I fear I may have underestimated the huge performance issues the production environment is currently facing.